### ISSUE TITLE
Proposals for improving errors

### ISSUE THREAD

**Issue by bassosimone** [measurement-kit/measurement-kit#935]
In this issue I shall summarize all the aspects of the current error systems that I don't like and will sketch out how I propose to fix them. Then I will also discuss things that I like and I'd expand.

# Annoying issues

## Error codes are explicit and updated manually

This is really annoying and error prone. The easy fix for this issue is to use enumerations. More specifically we can probably use the `enum class` construct. For example:

```C++
namespace mk {
namespace dns {

enum class ErrorCode {
    FORMAT_ERROR,
    SERVER_FAILED_ERROR,
    NOT_EXIST_ERROR,
    /* ... */
};

/* ... */

MK_DEFINE_ERR(MK_ERR_DNS(FORMAT_ERROR), FormatError, "dns_lookup_error")
MK_DEFINE_ERR(MK_ERR_DNS(SERVER_FAILED_ERROR), ServerFailedError, "dns_lookup_error")
MK_DEFINE_ERR(MK_ERR_DNS(NOT_EXIST_ERROR), NotExistError, "dns_lookup_error")

/* ... */
}}
```

## The category is encoded along with the error

Currently we encode error code and error category (e.g. `dns` or `net` together). I think it would be better to encode the category separately. The function to check for error equality could then become:

```C++
bool Error::operator==(Error &other) {
    return this->category == other.category && this->code == other.code;
}
```

This is a good place to mention that the current implementation of `Error` derives from `std::exception` because I believe it's handy to throw/catch errors sometimes. Given what I'm describing here, however, it would be worth investigating whether it would make sense to use [C++11 error_condition](http://en.cppreference.com/w/cpp/error/error_condition). It would also be worth investigating whether we can use not only the error condition but also the error category here.

## We do not map low-level errors

In most cases, when our code fails, we do not report the lowest-level error that occurred (for example, we see a `recv()` failing but we report something like "recv failed" rather than "host unreachable"). We can improve this situation by using the [portable errors defined by the standard C++11 library](http://en.cppreference.com/w/cpp/error/errc) that include many interesting errors including for example "host unreachable", "connection reset by peer", etc.

## No clear distinction between description and OONI errors

OONI defines errors as strings. We construct errors with macros where we specify the error code and the corresponding OONI error. We should extend this model to include, in addition to OONI error, a more descriptive explanation of the error itself. In many cases, OONI errors map to `undefined_error $code`, which is fine for OONI purposes but not that clear for the user.

## We often print errors as integers or print error codes

When we print errors as integers, we most often obtain `1` (the C++ equivalent of `True`) because errors are not convertible to integers but are convertible to booleans. If we print the error code, we obtain just a number that is not clear what it means. We should instead print the descriptive error (see previous point) using the logger functions (even better if we can extend logger as described in #868).

# Nice things

## Errors can be nested

I really like the feature that you can construct an error using another error to indicate that a certain level API failed and that the reason why it failed is a lower level error. For example, we can tell the user that we cannot negotiate with a NDT server because the function to receive the server version failed, and in turn this was caused by `recv()` failing, and in turn this was caused by the `std::errc::host_unreachable` C++11 portable error (equivalent of `EHOSTUNREACH` errno on POSIX systems).

This is also interesting because it allows us to group errors within a specific library. So, for example, the NDT test is composed of several phases (negotiating with server, upload test, download test, sending results to server). In turn, each phase is composed of many operations which can fail. Nesting errors would allow us to define &mdash; for the NDT library &mdash; generic errors such as "the download failed" and more specific errors such as "cannot connect to server". So, for example, I imagine that printing the error one would see the following information:

```
error: NDT test failed: network unreachable; more in detail:
- ndt::run_tests() failed; caused by
- ndt::test_s2c::run() failed; caused by
- ndt::test_s2c::coroutine() failed; caused by
- net::connect_many() failed; caused by
- net::connect() to "my.host.local" port 54321 failed; caused by
- net::connect_logic() failed; caused by
- net::connect_first_of() failed; caused by
- net::connect_base() to 1.2.3.4 port 54321 failed; caused by
- mk_bufferevent_on_event() failed; caused by
- system error: network unreachable
```

This can already be obtained with existing code (is similar to what `explain()` does) except that we need to define more strictly what errors to define. What I like of the above is that clearly tells one what problem has happened and aids a developer to quickly understand what went wrong.

## We can attach arbitrary context to errors

Basically, a error can contain a pointer to `ErrorContext`. This is useful when we want the caller to have much more information about a specific error and take further decisions beyond the chain of calls that failed and the specific error code returned by the system or by us.

**Comment by hellais** [comment]
Thanks for this excellent summary.

More comments inline:

> Error codes are explicit and updated manually
> This is really annoying and error prone. The easy fix for this issue is to use enumerations. More specifically we can probably use the `enum class` construct.

I actually quite like the fact that error codes are explicit and that by just looking at the code I can figure out what a specific error code maps to in terms of error string or meaning.

If this were to be lost by the fact that it's move into a macro/enumerator, there should at least be some tool that automatically generates docstrings with the error code and adds it to the code to preserve this ability.

> The category is encoded along with the error
> Currently we encode error code and error category (e.g. `dns` or `net` together). I think it would be better to encode the category separately. 

Yes, I believe this would be a useful addition. It seems like using something standard like `std::error_condition` is a good way to go.

> We do not map low-level errors

+1 to this. I think this would be very useful and is actually very important also for OONI to know what is the specific lower level error code.

> No clear distinction between description and OONI errors

Yes, this would be very useful too.

> We often print errors as integers or print error codes

Is this due to a limitation of the current way in which we consider error or is it just that we are not using them properly in the code?

I mean shouldn't you be able to always print the error_code from the code?
Maybe having some utility method or function that is used to print errors seems like the way to go.

> Errors can be nested

Yeah I also really like this.

In thinking about how to process them, though, we need to be careful in how we end up serialising the errors inside of the actual report to make it most useful to extra this information.

> We can attach arbitrary context to errors

Yeah, this quite solid as well. Nothing particular to comment on this point.

**Comment by bassosimone** [comment]
As an additional comment to this thread, I'd say that perhaps we should aim at having lots of parsing and network errors, with context, to measure anomalies, whereas we can collapse many more EDOOFUS-type errors into the same error and distinguish between them with an attribute string passed to the constructor. This seems like a sane balance between having few errors without being able to cover all cases and having too many errors. This occurred to me while discussing in privately several ooni-mobile-related issues with a user.

**Comment by bassosimone** [comment]
I have actually recently removed support for attaching arbitrary context to errors as part of #1261. The only use case we had for that feature could be implemented with less code and less hassle to access the data by moving the interesting bits of information in the `Transport` and by making sure that `connect()` always return some type of `Transport`, even on error. Since there were no users left of this feature, I decided to ditch it, in the interest of eliminating optional complexity.
